File: docs/database/migration/fix-string-literals.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""修复「字符串字面量被误伤」：把 SQL 单引号内的新表名还原成旧表名。

背景：全词替换 \\b<table>\\b 会命中 'model.models'、'/api/users'、
'https://openrouter.ai/models' 这类字面量（. / - 都是词边界），
导致菜单编码、API 路径绑定、外部 URL 被改坏。

用法：python fix-string-literals.py <mapping.csv> <target.sql> [...]
"""
import re
import sys

mapfile, files = sys.argv[1], sys.argv[2:]
if not mapfile or not files:
    print("usage: fix-string-literals.py mapping.csv file.sql [...]")
    sys.exit(1)

rev = {}
with open(mapfile, encoding="utf-8-sig") as fh:
    for line in fh:
        line = line.strip().replace("\r", "")
        if not line or line.startswith("#"):
            continue
        parts = line.split(",")
        if len(parts) < 2:
            continue
        old, new = parts[0].strip(), parts[1].strip()
        if old and new:
            rev[new] = old

keys = sorted(rev, key=len, reverse=True)  # 长名优先
pat = re.compile(r"('[^']*')")
subs = [(re.compile(r"\b" + re.escape(k) + r"\b"), rev[k]) for k in keys]


def fix(match):
    text = match.group(1)
    for regex, old in subs:
        text = regex.sub(old, text)
    return text


for path in files:
    with open(path, encoding="utf-8", newline="") as fh:
        content = fh.read()
    new = pat.sub(fix, content)
    if new != content:
        with open(path, "w", encoding="utf-8", newline="") as fh:
            fh.write(new)
        print("fixed:", path)
    else:
        print("unchanged:", path)

File: docs/database/migration/fix-string-literals.pl
#!/usr/bin/env perl
# 修复「字符串字面量被误伤」：把 SQL 单引号内的新表名还原成旧表名。
# 背景：全词替换 \b<table>\b 会命中 'model.models'、'/api/users'、'https://openrouter.ai/models'
#       这类字面量（. / - 都是词边界），导致菜单编码、API 路径、外部 URL 被改坏。
# 用法：perl fix-string-literals.pl <mapping.csv> <target.sql> [...]
use strict;
use warnings;

my ($mapfile, @files) = @ARGV;
die "usage: $0 mapping.csv file.sql [...]\n" unless $mapfile && @files;

open my $mh, '<', $mapfile or die "cannot read $mapfile: $!";
my %rev;
while (my $line = <$mh>) {
    chomp $line;
    $line =~ s/\r//g;
    next if $line =~ /^\s*#/;
    my ($old, $new) = split /,/, $line;
    next unless defined $old && defined $new && $old ne '' && $new ne '';
    $old =~ s/^\x{ef}\x{bb}\x{bf}//;
    $rev{$new} = $old;
}
close $mh;
# 长名优先，避免短名先命中
my @keys = sort { length($b) <=> length($a) } keys %rev;

for my $file (@files) {
    local $/;
    open my $fh, '<', $file or die "cannot read $file: $!";
    my $c = <$fh>;
    close $fh;
    my $before = $c;
    # 只处理单引号字面量内部
    $c =~ s{('(?:[^']*)')}{
        my $s = $1;
        for my $k (@keys) { $s =~ s/\b\Q$k\E\b/$rev{$k}/g; }
        $s;
    }ge;
    if ($c ne $before) {
        open my $out, '>', $file or die "cannot write $file: $!";
        print $out $c;
        close $out;
        print "fixed: $file\n";
    } else {
        print "unchanged: $file\n";
    }
}
